### ISSUE TITLE
Rule Suggestion: Warn on usage of .bind(this), prefer arrow func

### ISSUE THREAD

**Issue by SimenB** [eslint/eslint#3357]
Tried searching for it...

I'd like to get a linting warning on all usages of `function () {}.bind(this);`, as it could be replaced by `() => {}`. This is stylistic.

Warning: 
```js
$.ajax(/**/, {
  error: function (model, res) {
    this.someFuntion();
  }.bind(this)
})
```

OK: 
```js
$.ajax(/**/, {
  error: (model, res) => {
    this.someFuntion();
  }
})
```

Also OK:
```js
$.ajax(/**/, {
  error: this.someFuntion.bind(this)
})
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by BYK** [comment]
Related:

- #3140
- #3113
- #2800

**Comment by SimenB** [comment]
Yeah I found those. This is always safe though, as binding this is essentially the same thing, so you are _always_ safe with this change. Shouldn't be any need to check content or parents. But IDK how AST works, so I could be wrong :laughing:

**Comment by nzakas** [comment]
@SimenB I'd suggest starting out with a custom rule for this, as it seems a bit too specific.

**Comment by IanVS** [comment]
For the record, binding `this` is not what happens in arrow functions, which lexically scope `this` (it is essentially sugar for `_this = this` or `self = this`, etc.).

**Comment by michaelficarra** [comment]
@IanVS in what way is that different? Please provide an example.

**Comment by IanVS** [comment]
Here's an example from [YDKJS](https://github.com/getify/You-Dont-Know-JS/blob/master/this%20%26%20object%20prototypes/ch2.md#lexical-this):

```javascript
function foo() {
    // return an arrow function
    return (a) => {
        // `this` here is lexically adopted from `foo()`
        console.log( this.a );
    };
}

var obj1 = {
    a: 2
};

var obj2 = {
    a: 3
};

var bar = foo.call( obj1 );
bar.call( obj2 ); // 2, not 3!
```

**Comment by IanVS** [comment]
Thanks for challenging me, I think I was misunderstanding [YDKJS](https://github.com/getify/You-Dont-Know-JS/blob/master/this%20%26%20object%20prototypes/ch2.md#lexical-this).  These two cases seem identical:

```javascript
function foo() {
    // return an arrow function
    return (a) => {
        // `this` here is lexically adopted from `foo()`
        console.log( this.a );
    };
}

var obj1 = {
    a: 2
};

var obj2 = {
    a: 3
};

var bar = foo.call( obj1 );
bar.call( obj2 ); // 2, not 3!
```

is the same as: 

```javascript
function foo() {
    // return an bound function
    return function (a) {
        console.log( this.a );
    }.bind(this);
}

var obj1 = {
    a: 2
};

var obj2 = {
    a: 3
};

var bar = foo.call( obj1 );
bar.call( obj2 ); // 2, not 3!
```

I think what I missed is that `bind` itself does not avoid lexical scoping, but can be used in ways that do.  I guess I need to do some more studying on this.

**Comment by IanVS** [comment]
According to [this comment]( https://github.com/getify/You-Dont-Know-JS/issues/513#issuecomment-130300480), there are indeed some differences between arrow functions and `.bind(this)`:

> 1. Arrow functions are always anonymous, which means you can't, for instance, reliably call them recursively (since you don't have a reliable lexical name to use).
2. Arrow functions actually create lexical bindings for  `this` ,  `super` , and  `arguments` . Only  `this`  is bound by  `.bind(..)` .

With that in mind, would this rule still make sense?

**Comment by getify** [comment]
Arrow function also lexically binds `arguments` and `super`, which `bind(..)` does not. If either of those were present inside the function in question, warning about it needing to be an arrow would be a false signal, as that would alter its behavior. Also, if the function in question had a lexical name used for recursion purposes, changing it to an arrow would be inappropriate (as they are anonymous).

To illustrate, I wouldn't want false warnings in cases like this:

```js
var o1 = {
   id: "o1"   
};
var o2 = {
   id: "o2"
};

function foo() {
   setTimeout(function fn(){
      var total = 0;
      // check for any odd numbers as arguments, make them even
      for (var i=0; i<arguments.length; i++) {
         if (arguments[i] % 2 == 1) arguments[i]++;
         total += arguments[i];
      }
      if (total < 15) setTimeout( fn.bind(o2), 100, arguments[0]*2, arguments[1]*2 );
      else console.log(this.id,total);
   }.bind(this),100,2,4,7);
}

foo.call(o1);
// "o2" 24
```

**Comment by getify** [comment]
Another issue with blinding insisting on swapping `function(){}.bind(this)` for arrow: a bound function can be used in a `new` expression, but an arrow function cannot. So you'd need to be able to check to see how the function was being used to make sure this was a valid suggestion (warning) to make that wouldn't break the program.

**Comment by nzakas** [comment]
Good discussion. End result is still the same: this is better off as a custom rule.
